Add an absorbed power fraction to the generic heat source description #672

demoncoder-crypto opened the pull request:
1)- I tried to add the absorption_fraction parameter to GenericIonElHeatSourceConfig, RuntimeParams, RuntimeParamsProvider, and DynamicRuntimeParams in generic_ion_el_heat_source.py.

2)- Changed the calc_generic_heat_source function to use absorbed_power = Ptot * absorption_fraction as not all injected power is absorbed by the plasma. Some portion is lost through shine-through, mode conversion, or edge losses.

3)- Added the P_generic_injected field to the PostProcessedOutputs class in state.py

jcitrin commented on torax/sources/tests/generic_ion_el_heat_source_test.py:
```diff
@@ -28,6 +36,129 @@ def setUpClass(cls):
         model_func=generic_ion_el_heat_source.default_formula,
     )
 
+  def test_absorption_fraction(self):
```
This test could be made a lot simpler by directly unit testing calc_generic_heat_source , which has a simple API which does not necessitate building the full stack of TORAX data structures as done here.

jcitrin commented on torax/sources/tests/generic_ion_el_heat_source_test.py:
```diff
@@ -28,6 +36,129 @@ def setUpClass(cls):
+    
+    # Test that the absorbed power is scaled by absorption_fraction
+    # Profile 2 should have half the power of profile 1
+    integrated_ion1 = jax.scipy.integrate.trapezoid(ion1 * geo.volume, geo.rho)
```
just use:

integrated_ion1 = math_utils.volume_integration(ion1, geo)

same for all the other integrations.

This is how it's done in torax.post_processing. Take a look there: https://github.com/google-deepmind/torax/blob/main/torax/post_processing.py

jcitrin commented on torax/sources/tests/generic_ion_el_heat_source_test.py:
```diff
@@ -28,6 +36,129 @@ def setUpClass(cls):
+    integrated_el1 = jax.scipy.integrate.trapezoid(el1 * geo.volume, geo.rho)
+    integrated_el2 = jax.scipy.integrate.trapezoid(el2 * geo.volume, geo.rho)
+    
+    self.assertAlmostEqual(integrated_ion2 / integrated_ion1, 0.5, places=5)
```
use `np.testing.assert_allclose`  . And instead of `places=5` use `rtol=1e-5`

jcitrin replied:
please make this change. I think assertAlmostEqual will work too but we're trying to maintain a uniform convention

jcitrin commented on torax/post_processing.py:
```diff
@@ -60,13 +60,15 @@ def _calculate_integrated_sources(
     geo: geometry.Geometry,
     core_profiles: state.CoreProfiles,
     core_sources: source_profiles.SourceProfiles,
+    dynamic_runtime_params_slice=None,
```
do proper typing. Also, no need for default None

jcitrin commented on torax/post_processing.py:
```diff
@@ -121,6 +124,25 @@ def _calculate_integrated_sources(
         integrated['P_external_el'] += integrated[f'{value}_el']
+        
+        # Track injected power for generic heat source
+        if key == 'generic_ion_el_heat_source' and dynamic_runtime_params_slice is not None:
```
no need for the `and dynamic_runtime_params_slice is not None` . This should always be passed into this function. The None branch should never happen

jcitrin commented on torax/post_processing.py:
```diff
@@ -121,6 +124,25 @@ def _calculate_integrated_sources(
+        
+        # Track injected power for generic heat source
+        if key == 'generic_ion_el_heat_source' and dynamic_runtime_params_slice is not None:
+          from torax.sources import generic_ion_el_heat_source
```
Imports should be in the module header.

jcitrin commented on torax/post_processing.py:
```diff
@@ -121,6 +124,25 @@ def _calculate_integrated_sources(
+              integrated['P_generic_injected'] = integrated[f'{value}_tot']
+              integrated['P_external_injected'] += integrated['P_generic_injected']
+          except (KeyError, AttributeError):
+            integrated['P_generic_injected'] = integrated[f'{value}_tot']
```
if there is an exception then raise the KeyError. It means that there's a bug (e.g. in the naming of the key). The code flow should not continue here

demoncoder-crypto replied:
I will test this exhaustively and find out.

jcitrin commented on torax/post_processing.py:
```diff
@@ -121,6 +124,25 @@ def _calculate_integrated_sources(
+              integrated['P_generic_injected'] = integrated[f'{value}_tot'] / absorption_fraction
+              integrated['P_external_injected'] += integrated['P_generic_injected']
+            else:
+              integrated['P_generic_injected'] = integrated[f'{value}_tot']
```
an error should be raised in the else block. The expectation is that the source params is a generic_ion_el_heat_source.DynamicRuntimeParams. If it isn't then something went wrong

jcitrin commented on torax/post_processing.py:
```diff
@@ -121,6 +124,25 @@ def _calculate_integrated_sources(
+            source_params = dynamic_runtime_params_slice.sources[key]
+            if isinstance(source_params, generic_ion_el_heat_source.DynamicRuntimeParams):
+              # Calculate injected power from absorbed power and absorption fraction
+              absorption_fraction = jnp.maximum(source_params.absorption_fraction, constants.CONSTANTS.eps)
```
no need for this jnp maximum. Remove the line

demoncoder-crypto replied:
I thought for this one making sure that absorption fraction is never below some small value so preventing division by zero. I will remove this.

jcitrin replied:
No worries, it's good instinct to do so. The reason it's not needed is that all validation will be handled by pydantic. However, we are in the middle of pydantic development and in the midst of defining validations for TimeVaryingScalars. See https://github.com/google-deepmind/torax/pull/817/ . For absorption_fraction we can later define an appropriate validator in pydantic. If you like you can write a TODO above `absorption_fraction` in `GenericIonElHeatSourceConfig` to include later an appropriate pydantic validation when ready. We'll do it later.

jcitrin commented on torax/post_processing.py:
```diff
@@ -190,13 +213,16 @@ def make_outputs(
       geo,
       sim_state.core_profiles,
       sim_state.core_sources,
+      sim_state.dynamic_runtime_params_slice,
```
ToraxSimState does not have this attribute. Did you run pytest on TORAX or try to run a simple case? You would have seen runtime errors and would have caught this error. Please do so in the future to avoid extra work in reviews. 

Instead, you can pass a `DynamicRuntimeParamsSlice` into `make_output` where it is called. The appropriate object is indeed available in the calling function.

demoncoder-crypto replied:
For this i thought storing them in the state object ensures that the simulation state is fully self-contained and reproducible.

jcitrin replied:
`sim_state` is an input argument into `make_outputs` here. The type is a `state.ToraxSimState` , so one can look at the dataclass and see which attributes are available.

In general it is strongly recommend after making any changes to run the test (using pytest) of any part of TORAX that you are changing, and/or to run a basic case with `python3 run_simulation_main.py --config='torax.examples.basic_config'` just as a sanity test to make sure that nothing fundamental is broken. You will catch many issues this way. You can use standalone pytest or the GitHub CI on your branch (before making a PR) to see if tests pass first. Also, failing tests give hints on what went wrong. See also https://torax.readthedocs.io/en/latest/contribution_tips.html#contribution-tips

demoncoder-crypto replied:
Understood

demoncoder-crypto replied:
Thanks a lot, for such in depth feedback means a lot

jcitrin reviewed (commented):
Thanks for the work here. In addition to the changes requested in the comments made, I suggest to also add an absorption fraction to ion_cyclotron_source and do similar logic as done for generic_ion_el_heat_source. Apologies that it wasn't written as such in the issue, but those are the two source models where currently this feature would be very useful.

jcitrin commented:
Please sign the CLA such that this PR can eventually be merged

demoncoder-crypto commented:
I will work on this immediately. I have Signed the CLA I don't know why checks aren't confirming them.

jcitrin commented:
> I will work on this immediately. I have Signed the CLA I don't know why checks aren't confirming them.

see comment in the other PR

demoncoder-crypto commented on torax/sim.py:
Understood.

demoncoder-crypto commented:
The changes- 

The make_outputs function in post_processing.py accepts a dynamic_runtime_params_slice parameter, which it passes to calculate_integrated_sources. In calculate_integrated_sources, the dynamic_runtime_params_slice is reaching source parameters for heating sources such as generic_ion_el_heat_source and ion_cyclotron_source, they get to access the absorption_fraction attribute. The finalize_output method in step_function.py calls make_outputs and should pass dynamic_runtime_params_slice_t_plus_dt to it. The run_simulation function in sim.py also calls make_outputs and should pass dynamic_runtime_params_slice to it.

The test in post_processing.py passes self.dynamic_runtime_params_slice to make_outputs.

jcitrin commented on torax/post_processing.py:
```diff
@@ -125,22 +140,36 @@ def _calculate_integrated_sources(
+            if absorption_fraction < 1.0:
+              injected_power = total_absorbed / absorption_fraction
+              integrated['P_external_injected'] += injected_power
             else:
```
This can all be made much simpler now that the only now `post_processed_output` attribute is `P_external_injected` (which is a good change btw). 

Just do

`if source_params and hasattr(source_params, 'absorption_fraction'):` 

and if True, do
total_absorbed = integrated[f'{value}_ion'] + integrated[f'{value}_el']
injected_power = total_absorbed / absorption_fraction
integrated['P_external_injected'] += injected_power

and if False, do
integrated['P_external_injected'] += integrated[f'{value}_ion'] + integrated[f'{value}_el']


That's it. No need for all the TypeError checks and so on. 

Also no need to check absorption_fraction < 1.0 . Pydantic will take care of this

jcitrin commented on torax/post_processing.py:
```diff
@@ -173,22 +202,21 @@ def make_outputs(
     sim_state: state.ToraxSimState,
     geo: geometry.Geometry,
     previous_sim_state: state.ToraxSimState | None = None,
+    dynamic_runtime_params_slice: runtime_params_slice.DynamicRuntimeParamsSlice | None = None,
 ) -> state.ToraxSimState:
-  """Calculates post-processed outputs based on the latest state.
+  """Makes post-processed outputs given a simulation state.
 
-  Called at the beginning and end of each `sim.run_simulation` step.
   Args:
-    sim_state: The state to add outputs to.
-    geo: Geometry object
-    previous_sim_state: The previous state, used to calculate cumulative
-      quantities. Optional input. If None, then cumulative quantities are set at
-      the initialized values in sim_state itself. This is used for the first
-      time step of a the simulation. The initialized values are zero for a clean
-      simulation, or the last value of the previous simulation for a restarted
-      simulation.
+    sim_state: The simulation state to post-process.
+    geo: Magnetic equilibrium geometry.
+    previous_sim_state: The previous simulation state. If provided, used to
+      calculate time derivative of time-resolved quantities.
+    dynamic_runtime_params_slice: Optional dynamic runtime parameters that may
+      be used for power calculations. If not provided, injected power will be
+      equal to absorbed power.
 
   Returns:
-    sim_state: A ToraxSimState object, with any updated attributes.
+    A copy of the input simulation state with post-processed outputs.
```
Do not change the docstring.

This change made the docstring  _less_  accurate

demoncoder-crypto replied:
Understood.

jcitrin commented on torax/state.py:
Please revert all changes to this file apart from the `P_external_injected` additions. They are not necessary changes for this PR and also many of the changes lead to incorrect or insufficient descriptions

demoncoder-crypto replied:
Do you want me to revert all of the changes and just keep p_external_injected?

jcitrin replied:
yes

demoncoder-crypto replied:
Understood.

jcitrin commented on torax/state.py:
```diff
@@ -488,7 +486,6 @@ def zeros(cls, geo: geometry.Geometry) -> PostProcessedOutputs:
         q95=jnp.array(0.0),
         Wpol=jnp.array(0.0),
         li3=jnp.array(0.0),
-        P_generic_injected=jnp.array(0.0),
```
fine to remove this but you need to add instead `P_external_injected` 

Please follow my previous recommendations on running the simple TORAX case and the TORAX tests. That would have caught this error before review

jcitrin commented on torax/tests/post_processing.py:
```diff
@@ -80,31 +81,36 @@ def setUp(self):
 
   def test_make_outputs(self):
-    """Test that post-processing outputs are added to the state."""
+    """Checks that outputs are made correctly."""
```
do not change the docstring

jcitrin commented on torax/tests/post_processing.py:
```diff
@@ -80,31 +81,36 @@ def setUp(self):
   def test_make_outputs(self):
-    """Test that post-processing outputs are added to the state."""
+    """Checks that outputs are made correctly."""
+    self.core_profiles.j_phi = jax.jit(formulas.calc_current_density)(
+        p=self.core_profiles.pressure.cell_value(),
+        psi=self.core_profiles.psi.face_value(),
+        rho=self.geo.rho,
+        r=self.geo.r,
+        r_vol=self.geo.r_vol,
+        vol=self.geo.volume,
+    )
```
why is this added? It doesn't make sense. j_phi is not a core_profiles attribute and anyway is not needed here.

jcitrin commented on torax/tests/post_processing.py:
```diff
@@ -80,31 +81,36 @@ def setUp(self):
-        t=jax.numpy.array(0.0),
-        dt=jax.numpy.array(0.1),
-        time_step_calculator_state=None,
+        transport_coefficients=state.TransportCoefficients.zeros(self.geo),
         post_processed_outputs=state.PostProcessedOutputs.zeros(self.geo),
-        stepper_numeric_outputs=state.StepperNumericOutputs(
-            outer_stepper_iterations=1,
-            stepper_error_state=1,
-            inner_solver_iterations=1,
-        ),
-        geometry=self.geo,
     )
```
no need to change the sim_state assignment

jcitrin commented on torax/tests/post_processing.py:
```diff
@@ -128,6 +134,7 @@ def test_calculate_integrated_sources(self):
+        self.dynamic_runtime_params_slice,
     )
     # pylint: enable=protected-access
 
```
you will need to extend `expected_keys` with `P_external_injected`

jcitrin commented:
Forgot to mention. 
1. Need to merge this branch with TORAX head to resolve merge conflicts
2. Additionally, there is a missing `dynamic_runtime_params` argument input into `post_processing.make_outputs` in `orchestration/step_function` . This missing argument would also have been caught by trying to run TORAX or run the tests, so please do so ahead of the next commit.

demoncoder-crypto commented:
Changes implemented:
Ffixed most of the import issues, I ran both the tests and both of them were successful.
Updated the test_make_outputs function to match the current structure of the ToraxSimState class. The only one I am unsure about is using  jax.numpy.array() for the t and dt value. As I understand JAX is strict about types so using plain Python floats (1.0, 0.1) is generally discouraged but I am not sure about this.
P_external_injected to the expected_keys set in the test_calculate_integrated_sources function, which now passes.
Removed the unnecessary j_phi calculation from the test_make_outputs function.
All the tests in the PostProcessingTest class are now passing, which means our changes to add the absorption_fraction feature and track injected power are working correctly.

demoncoder-crypto commented:
I will work on CLA just right after all of the fixes. I will create documentation as requested right now. The tests are failing because P_external_injected-- i will fix immediately. Also thanks a lot for bearing with me and helping me so much.

jcitrin commented on docs/output.rst:
```diff
@@ -207,6 +207,12 @@ post_processed_outputs
 These outputs are calculated by the post_processing module, for both
 analysis and inspection.
 
+The post-processing module uses the ``dynamic_runtime_params_slice`` parameter to 
+calculate various power-related outputs. This parameter provides access to runtime 
+parameters that may change during simulation, such as absorption fractions for 
+different heating sources. This ensures accurate calculation of power flows, 
+particularly for outputs like ``P_external_injected`` and fusion gain (``Q_fusion``).
+
```
Revert this block. Not needed

jcitrin replied:
Please do this

jcitrin commented on docs/configuration.rst:
```diff
@@ -1498,6 +1498,17 @@ We envisage this feature being useful for example to:
-  a simulation up to a certain point in time. After running the initial
-  simulation you could then modify and sweep the runtime parameter config in
-  order to do some uncertainty quantification.
+  a simulation you could then modify and sweep the runtime parameter config in
+  different ways.
+
+Dynamic Runtime Parameters in Post-Processing
+=============================================
+
+The post-processing module uses dynamic runtime parameters to calculate various outputs, 
+particularly for power calculations. The ``dynamic_runtime_params_slice`` parameter is 
+passed to the ``make_outputs`` function to ensure that power calculations correctly 
+account for parameters that may change during simulation, used in absorption fractions 
+for different heating sources.
+
+This is particularly important for calculations like fusion gain (Q_fusion), which 
+depends on the external injected power before absorption effects.
```
Revert this change. Not needed.

The only change that's needed is to add an explanation of the new `absorption_fraction` variable to both the `ion_cyclotron_source` and `generic_ion_el_heat_source` sections

demoncoder-crypto replied:
On it

jcitrin commented on torax/sources/generic_ion_el_heat_source.py:
```diff
@@ -111,6 +121,7 @@ def calc_generic_heat_source(
     w: float,
     Ptot: float,
     el_heat_fraction: float,
+    absorption_fraction: float = 1.0,
```
no need for the default value here. The config system will already provide the default value for absorption_fraction if nothing else is set

jcitrin commented on torax/sources/source_models.py:
The changes to this file should be reverted. It belongs to a different PR

demoncoder-crypto replied:
The documentation changes? I didnt change Source_model.py. I have kept it aside for a sperate issue. let me check once again

jcitrin replied:
yeah the changes to source_model.py corresponding to the separate issue on Mavrin+Bremsstrahlung are showing up here

demoncoder-crypto replied:
Omg I am so sorry let me clean up. I was keeping 2 folders somehow I staged local changes in both of them.

jcitrin replied:
No worries

demoncoder-crypto replied:
I have committed Some changes if they are fixed let me know if any other changes are needed i will fix them immediately

jcitrin commented on torax/sources/tests/source_models_validation_test.py:
please carry this out

demoncoder-crypto replied:
Yes I am removing all of the changes and keeping a sperate folder for #821 somehow i messed them up. Sorry

jcitrin commented:
A couple of remaining issues:

1. Some tests are failing because of the new output variable. See "important" at the bottom of https://torax.readthedocs.io/en/latest/contribution_tips.html#contribution-tips . This causes some tests to fail which are comparing the entire output xarray structures. To fix this, follow the TORAX instructions on running a simulation https://torax.readthedocs.io/en/latest/running.html and run the following 3 config modules: `torax.test.test_data.test_implicit`, `torax.test.test_data.test_changing_config_before`, and `torax.test.test_data.test_changing_config_after`. After each successful run, a data file will be written to a `/tmp/<my run>/state_history.nc` directory (the exact directory name will be printed after simulation completion). Copy each file to `torax/tests/test_data/test_implicit.nc`, `torax/tests/test_data/test_changing_config_before.nc`, and `torax/tests/test_data/test_changing_config_after.nc` respectively. That will fix those tests. Sorry about this one, it's cryptic and I see now that the documentation is clear enough (will improve it).

2. The new ion_cyclotron_source test is broken. Please look at the error messages and fix

jcitrin commented on torax/sources/tests/source_models_validation_test.py:
revert all changes to this file (related to the Mavrin + Bremsstrahlung issue)

demoncoder-crypto replied:
I have tried to revert changes for Mavrin + Bremsstrahlung issue. And I have ran all tests. Check for any issues i will fix them immediately

jcitrin commented:
OK the big breaking changes already came in :) https://github.com/google-deepmind/torax/pull/826

This PR will need to be synced to head and the merge conflicts fixed.

Sorry about that. We are in the middle of a major config patterns and output refactor ahead of a new release.

demoncoder-crypto commented:
![Uploading Screenshot 2025-03-12 165732.jpg…]()

demoncoder-crypto commented:
![Screenshot 2025-03-12 165732](https://github.com/user-attachments/assets/474b6a8f-8646-4fdc-ac7d-5bfa83e05f3a)
This is the problem I am facing although registered it says issues there

demoncoder-crypto commented:
I will sync to head and resolve the merge conflicts.

demoncoder-crypto commented:
The tests in sim_test.py are failing because can't find the QLKNN model.

demoncoder-crypto commented:
I have Pushed at gh pr checkout 831

jcitrin commented:
About the CLA. I think it may possibly be due to a mismatch in the email address provided in the CLA, and the email address in the git commits.

If I clone your fork and check out your branch, and do `git log`, I see your email address has a space in it, which may be causing the mismatch.

I recommend to change your email address associated with your commits with `git config` (look online for the exact command to use). 

You then need to fix your existing commits on the PR. I think there are two options:
1. Retroactively changing the email address used in your previous commits (will need to change the commit history). Squashing same commits may help.
2. (This is a hack), add your "wrong" email with the space as an alternative email to the CLA.

demoncoder-crypto commented:
I will just do the Hack and after this works I will fix the whole thing

demoncoder-crypto commented:
Hey does it work. Or do i need to make some changes on my end?

jcitrin commented:
> Hey does it work. Or do i need to make some changes on my end?

I'm not sure what you did but the CLA still doesn't pass.

Try doing the first option. Correct your email address in git config on the command line:

`git config --global user.email "dhillonprabhjitsingh@gmail.com"`

Then amend your commits. For example for amending the last commit:

`git commit --amend --no-edit --reset-author` 

But you'd need to amend all your commits in the PR. 
1. You can try (e.g. with rebase) squashing your previous commits into one commit and ensure that that single commit has the correct email
2. You can make a new branch `git checkout <newbranch>` and then squash merge this old branch in the PR `git merge --squash <oldbranch>` and send a new PR.  Replace <newbranch> and <oldbranch>  with the appropriate branch names

demoncoder-crypto commented:
![Screenshot 2025-03-13 111430](https://github.com/user-attachments/assets/a5a37098-c696-490a-8dc0-6c2344e6f659)

demoncoder-crypto commented:
The checks passed at #832 for CLA as it wasn't passing here
